# My choice:

Instead of following the instructors and doing the lesson in Google Colab, I preferred Jupyter, since I already had it installed from the "Hashtag" course, a few weeks ago.

# What I did on May 24, 2022:

### 1) Imported pandas "as" pd (a nickname)
### 2) Imported data from gist and at the same time creatind a variable "url" containing the imported data.
### 3) Named "dados" the variable containing the csv result of "url".
### 4) Gave the command "head", which shows:

a- The first five entries of the Data Frame (DF) if you leave the ".head" parenthesis empty;

b- The "n" entries if you fill the ".head" parenthesis with a number.

# I had to stop and I'm continuing the morning after.

Ok, let's do the lesson!

### Note that the code cells here are re-executable.

You may change a parameter and run again.

A different result will display in place of the former one.

### I guess everytime you enter here you must run all cells again to re-arrange the setup.

In [1]:
import pandas as pd
url = "https://gist.githubusercontent.com/tgcsantos/3bdb29eba6ce391e90df2b72205ba891/raw/22fa920e80c9fa209a9fccc8b52d74cc95d1599b/dados_imoveis.csv"
dados = pd.read_csv(url)
dados.head()

Rua              Bairro     Cidade  Metragem  \
0           Avenida Itacira, 255    Planalto Paulista  São Paulo      1000   
1  Rua Aurelia Perez Alvarez, 42   Jardim dos Estados  São Paulo       496   
2                 Rua Alba Valdez     Jardim Reimberg  São Paulo       125   
3                             NaN      Jardim Morumbi  São Paulo       310   
4        Rua Tobias Barreto, 195                Mooca  São Paulo       100   

   Quartos  Banheiros  Vagas         Valor  
0        4          8      6  R$ 7.000.000  
1        4          4      4  R$ 3.700.000  
2        4          3      2    R$ 380.000  
3        3          2      4    R$ 685.000  
4        3          2      2    R$ 540.000

# Explaining the code above:

### "url=" means:

1) creating a variable called "url" (could be any other name, to avoid confusion with its origin);

2) the variable content is "called" from an url. Something that is at that address;


### "dados":


1) I know that url contains a raw tabular data = csv (The CSV, Comma Separated Values, file is a text file with a specific format to enable saving the data in a table-structured format.)

2) A variable named "dados" contains csv data from another just created variable called "url";

3) "dados" contains the Pandas reading of the imported "url" variable's tabular data.

### "dados.head()" displays:

a- The first five entries of the Data Frame (DF);

b- The "n" entries if you fill the ".head" parenthesis with a number. You may try it in the above code.

### One of the teachers  always commands "head" first thing to check if  the DF was imported ok.


# knowing more about the just imported DF

Just type the variable name that containd the imported DF: "dados", as below.

The DF will be displayed: the 1st five lines and the last 5.

After that an important info is given: "10008 rows × 8 columns" = the size of the table.

In [2]:
dados

Rua              Bairro     Cidade  \
0               Avenida Itacira, 255    Planalto Paulista  São Paulo   
1      Rua Aurelia Perez Alvarez, 42   Jardim dos Estados  São Paulo   
2                     Rua Alba Valdez     Jardim Reimberg  São Paulo   
3                                 NaN      Jardim Morumbi  São Paulo   
4            Rua Tobias Barreto, 195                Mooca  São Paulo   
...                               ...                 ...        ...   
10003                      Rua Tapuçu          Vila Sofia  São Paulo   
10004                   Rua Guararema     Bosque da Saúde  São Paulo   
10005               Rua Estero Belaco       Vila da Saúde  São Paulo   
10006         Rua Manuel Onha, , 514        Vila Oratório  São Paulo   
10007      Rua Barão de Tramandaí, 0          Vila Alpina  São Paulo   

       Metragem  Quartos  Banheiros  Vagas  \
0          1000        4          8      6   
1           496        4          4      4   
2           125        4          3      2   
3           310        3          2      4   
4           100        3          2      2   
...         ...      ...        ...    ...   
10003        95        2          2      2   
10004       605        8          7      5   
10005       205        3          3      6   
10006       185        3          2      2   
10007       122        3          2      2   

                                      Valor  
0                              R$ 7.000.000  
1                              R$ 3.700.000  
2                                R$ 380.000  
3                                R$ 685.000  
4                                R$ 540.000  
...                                     ...  
10003                            R$ 665.000  
10004                          R$ 2.300.000  
10005                          R$ 1.050.000  
10006  R$ 2.200\n                      /Mês  
10007                            R$ 270.000  

[10008 rows x 8 columns]

### From the 2 displayed data above we can see that there are things to be cleaned/corrected in the DF:

1) Missing street names;

2) Currency symbols (all lines) and words (line 10006 at least) mixed with numbers at the "Valor" column;

3) Double comma in line 10006's address;

4) Whatever comes next.

### If I want to see more from the DF:

"dados.sample()" displays 1 RANDOM line.

By filling the parenthesis with a number, that quantity of random lines will be displayed.

In [3]:
dados.sample()

Rua                 Bairro     Cidade  Metragem  Quartos  \
5386  Rua João Jabotam  Vila Anglo Brasileira  São Paulo       272        3   

      Banheiros  Vagas                                 Valor  
5386          2      2  R$ 6.200\n                      /Mês

In [4]:
dados.sample(15)

Rua             Bairro     Cidade  Metragem  \
3858           Rua Marquês Maricá             Sacomã  São Paulo       171   
6647       Rua Cardoso de Almeida           Perdizes  São Paulo       350   
7365    Rua Marechal Bina Machado   Jardim Marajoara  São Paulo       450   
7100   Rua Evaristo da Veiga, 81             Catumbi  São Paulo       125   
5179                 Rua Apinajés             Sumaré  São Paulo       140   
7660  Rua Alphonsus de Guimaraens            Santana  São Paulo       255   
9297                          NaN        Vila Romana  São Paulo       169   
7759           Rua Manuel Jacinto         Vila Morse  São Paulo       261   
9931                   Rua Tapuçu         Vila Sofia  São Paulo        90   
1566              Rua dos Tamanás      Vila Madalena  São Paulo       305   
5313                   Rua Epeira  Alto de Pinheiros  São Paulo       195   
9870     Rua Brasileiro de Campos       Vila Nhocune  São Paulo       145   
6006                          NaN      Jardim Europa  São Paulo       172   
9733          Rua Moreira Cardoso   Jardim Aeroporto  São Paulo       220   
644                   Rua Araporé     Jardim Guedala  São Paulo       626   

      Quartos  Banheiros  Vagas                                 Valor  
3858        3          3      2                            R$ 563.000  
6647        5          4      5  R$ 7.000\n                      /Mês  
7365        3          2      4                          R$ 2.500.000  
7100        2          3      2                            R$ 749.000  
5179        3          2      1                          R$ 1.380.000  
7660        2          3      5                          R$ 1.200.000  
9297        3          3      2                          R$ 1.100.000  
7759        3          3      2                            R$ 800.000  
9931        2          2      2                            R$ 665.000  
1566        3          6      4                          R$ 3.500.000  
5313        3          2      2                          R$ 1.950.000  
9870        3          2      2                            R$ 450.000  
6006        2          3      2                          R$ 4.400.000  
9733        3          3      5                          R$ 1.100.000  
644         4          6      4                          R$ 4.200.000

# To get a "series":
Series means:

1) "A Pandas Series is like a column in a table. It is a one-dimensional array holding data of any type".

2) "Series is a one-dimensional labeled array capable of holding data of any type (integer, string, float, python objects, etc.). The axis labels are collectively called index".

Check how to get it below: (be careful, it's case sensitive)

In [5]:
dados["Bairro"]

0         Planalto Paulista
1        Jardim dos Estados
2           Jardim Reimberg
3            Jardim Morumbi
4                     Mooca
                ...        
10003            Vila Sofia
10004       Bosque da Saúde
10005         Vila da Saúde
10006         Vila Oratório
10007           Vila Alpina
Name: Bairro, Length: 10008, dtype: object

Ok, there are 10008 rows = the total given by the "dados" command above.

But I bet the same does not happen with the Rua column, because there are data missing or typed wrongly.

Let`s check:

In [6]:
dados["Rua"]

0                 Avenida Itacira, 255 
1        Rua Aurelia Perez Alvarez, 42 
2                       Rua Alba Valdez
3                                   NaN
4              Rua Tobias Barreto, 195 
                      ...              
10003                        Rua Tapuçu
10004                     Rua Guararema
10005                 Rua Estero Belaco
10006           Rua Manuel Onha, , 514 
10007        Rua Barão de Tramandaí, 0 
Name: Rua, Length: 10008, dtype: object

### Counting the "NaN" in "Rua":

In [7]:
dados['Rua'].isna().sum()

3434

In [8]:
3434/10008*100

34.31254996003197

So 34% of the DF does not have a street name. 1/3 is very significative!

What to do? Simply disregard? Erase them? Is the street nasme important or only the borrough matters?

Leave the streets there, I say. No harm so far.

Following the lesson:

What if I want to know the borrough of a certain row?

In [9]:
dados["Bairro"][6522]

'Vila Paulo Silas'

Done!

With "dados.info" I get what kind of data is in each column.

In [10]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10008 entries, 0 to 10007
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Rua        6574 non-null   object
 1   Bairro     9996 non-null   object
 2   Cidade     10008 non-null  object
 3   Metragem   10008 non-null  int64 
 4   Quartos    10008 non-null  int64 
 5   Banheiros  10008 non-null  int64 
 6   Vagas      10008 non-null  int64 
 7   Valor      10008 non-null  object
dtypes: int64(4), object(4)
memory usage: 625.6+ KB


In the "Dtype" column: Object means text and int64 means an integer number.

In the "Non-Null Count" column: how many valid data. In number of rows. Remember we have 10008 rows here.

So Rua and Bairro are incomplete.

And Valor has not the desired "int64" (integer numbers) kind of data. How to fix it? (later...)

To display a column we can command "dados["Rua"]"

as we did before...

OR...

"dados.Rua" (If the column name is one word (simple) the SAME result could ALSO be gotten this way)

Let's try:

In [11]:
dados.Rua

0                 Avenida Itacira, 255 
1        Rua Aurelia Perez Alvarez, 42 
2                       Rua Alba Valdez
3                                   NaN
4              Rua Tobias Barreto, 195 
                      ...              
10003                        Rua Tapuçu
10004                     Rua Guararema
10005                 Rua Estero Belaco
10006           Rua Manuel Onha, , 514 
10007        Rua Barão de Tramandaí, 0 
Name: Rua, Length: 10008, dtype: object

Ok.
So let`s calculate the average square meters using that:

In [12]:
dados.Metragem.mean()

442.30065947242207

So the mean square area of the properties is 442.3 square meters.

In [13]:
dados["Metragem"].mean()

442.30065947242207

If you want to take a look at a bunch of "square meters", just take the "mean" off the above code, like that:

In [14]:
dados["Metragem"]

0        1000
1         496
2         125
3         310
4         100
         ... 
10003      95
10004     605
10005     205
10006     185
10007     122
Name: Metragem, Length: 10008, dtype: int64

My remark:

You see? It looks like we have rather large apartments or houses here. Does that mean high income rate by household?

Another observation is that all the rows in this series are dully filled.

Continuing on May 26.

My remark:

### If there is any text cell that should be "markdown" running as "code", the "run all" stops at the error.

So watch out! Every new cell is born as "code".

If you want text, change it in the header's due option box.

### What is the average square meters in Vila Mariana?

1) Check if there is Vila Mariana in the DF;

2) just to paractice: If yes, check how many (one thing at a time, not a function).

In [15]:
dados["Bairro"] == "Vila Mariana"

0        False
1        False
2        False
3        False
4        False
         ...  
10003    False
10004    False
10005    False
10006    False
10007    False
Name: Bairro, Length: 10008, dtype: bool

In [16]:
(dados["Bairro"] == "Vila Mariana") *1

0        0
1        0
2        0
3        0
4        0
        ..
10003    0
10004    0
10005    0
10006    0
10007    0
Name: Bairro, Length: 10008, dtype: int32

In [17]:
sum((dados["Bairro"] == "Vila Mariana") *1)

184

Ok, now we see there are 184 properties in Vila Mariana.

# No one told us if they are apartments or houses

But there is a "Vagas" column. That means apartments with "n" places to park per apartment.

When dealing with houses they usually don't talk about "Vagas", places to park).

Continuing...

In [19]:
tem_imoveis_vila =(dados["Bairro"] == "Vila Mariana")

In [20]:
tem_imoveis_vila

0        False
1        False
2        False
3        False
4        False
         ...  
10003    False
10004    False
10005    False
10006    False
10007    False
Name: Bairro, Length: 10008, dtype: bool

Now we crop the "tem_imoveis_vila" from "dados":

In [21]:
dados[tem_imoveis_vila]

Rua        Bairro     Cidade  Metragem  \
100               Rua Pero Correia  Vila Mariana  São Paulo       250   
120   Praça Doutor Carvalho Franco  Vila Mariana  São Paulo       250   
155     Rua Professor João Marinho  Vila Mariana  São Paulo       170   
160                            NaN  Vila Mariana  São Paulo       228   
334         Rua Pero Correia, 892   Vila Mariana  São Paulo       350   
...                            ...           ...        ...       ...   
9645               Rua Sousa Ramos  Vila Mariana  São Paulo       285   
9701             Rua Benito Juarez  Vila Mariana  São Paulo       171   
9737             Rua Benito Juarez  Vila Mariana  São Paulo       171   
9807                           NaN  Vila Mariana  São Paulo       275   
9852              Rua França Pinto  Vila Mariana  São Paulo       280   

      Quartos  Banheiros  Vagas         Valor  
100         3          3      1  R$ 1.350.000  
120         3          5      2  R$ 2.450.000  
155         3          4      2  R$ 1.490.000  
160         4          5      3  R$ 2.200.000  
334         6          4     10  R$ 1.700.000  
...       ...        ...    ...           ...  
9645        3          5      5  R$ 1.650.000  
9701        3          2      4  R$ 1.650.000  
9737        3          2      4  R$ 1.650.000  
9807        3          1      6  R$ 1.500.000  
9852        3          3      4  R$ 1.600.000  

[184 rows x 8 columns]

Now we put this result in a variable:

In [22]:
imoveis_na_vila_mariana =dados[tem_imoveis_vila]

Done. From this variable, that now became a smaller DF, we request the square meter average.

In [23]:
imoveis_na_vila_mariana["Metragem"].mean()

233.6195652173913

Ready!

Comparing the Vila average area x the whole DF average area: (my own idea)

In [25]:
(imoveis_na_vila_mariana["Metragem"].mean() / dados.Metragem.mean())*100

52.819176325907726

Ok! The average area in Vila Mariana is 52.819176325907726 per cent of the total average.

It's a large apartment for Sao Paulo standard, though. Both figures are, in fact.